libsawtooth/src/publisher/factory.rs

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2022 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+//! Conatains the struct for starting publishing of a new artifact
+
+use std::sync::mpsc::channel;
+use std::thread;
+
+use crate::artifact::{Artifact, ArtifactCreator, ArtifactCreatorFactory};
+use crate::error::InternalError;
+
+use super::message::PublishMessage;
+use super::{
+    Batch, BatchExecutionResult, BatchVerifierFactory, PendingBatches, PublishHandle,
+    PublisherContext,
+};
+
+/// Factory to start publishing a new artificat
+///
+/// The `PublishFactory` starts up the exeuction and verification of batches by
+/// spinning up a thread and returning a `PublishHandle` that can be used to
+/// control the thread execution.
+///
+// allow type complexity here because of the use of associated types
+#[allow(clippy::type_complexity)]
+pub struct PublishFactory<
+    B: 'static + Batch,
+    C: 'static + PublisherContext,
+    R: 'static + Artifact,
+    I: 'static + BatchExecutionResult,
+> {
+    artifact_creator_factory: Box<
+        dyn ArtifactCreatorFactory<
+            ArtifactCreator = Box<dyn ArtifactCreator<Context = C, Input = Vec<I>, Artifact = R>>,
+        >,
+    >,
+    batch_verifier_factory:
+        Box<dyn BatchVerifierFactory<Batch = B, Context = C, ExecutionResult = I>>,
+}
+
+/// Create a new `PublishFactory`
+///
+/// Arguments
+///
+/// * `artifact_creator_factory` - a factory for creating a new artifact creator for a new
+///    publishing thread
+/// * `batch_verifier_factory` - a factory for creating a new batch verifier for a new publishing
+///    thread
+///
+// allow type complexity here because of the use of associated types
+#[allow(clippy::type_complexity)]
+impl<
```

Comment:
Generics constraints are far easier to read in a `where` clause

Reply:
Updated